gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java

```diff
@@ -105,60 +105,125 @@
   private final FileSystem fs;
 
   private volatile Job job;
-  private volatile boolean isCancelled = false;
+  private volatile Path mrJobDir;
 
-  public MRJobLauncher(Properties properties)
+  public MRJobLauncher(Properties properties, Properties jobProps)
       throws Exception {
-    this(properties, new Configuration());
+    this(properties, jobProps, new Configuration());
   }
 
-  public MRJobLauncher(Properties properties, Configuration conf)
+  public MRJobLauncher(Properties properties, Properties jobProps, Configuration conf)
       throws Exception {
-    super(properties);
+    super(properties, jobProps);
     this.conf = conf;
     URI fsUri = URI.create(this.properties.getProperty(ConfigurationKeys.FS_URI_KEY, ConfigurationKeys.LOCAL_FS_URI));
     this.fs = FileSystem.get(fsUri, conf);
   }
 
   @Override
-  public void cancelJob(Properties jobProps)
+  public void cancelJob(JobListener jobListener)
       throws JobException {
-    String jobName = jobProps.getProperty(ConfigurationKeys.JOB_NAME_KEY);
-
-    if (this.isCancelled || !Optional.fromNullable(this.job).isPresent()) {
-      LOG.info(String.format("Job %s has already been cancelled or has not started yet", jobName));
-      return;
-    }
-
     try {
-      // Kill the Hadoop MR if has not completed yet
-      if (!this.job.isComplete()) {
+      if (!this.isCancelled && this.job != null && !this.job.isComplete()) {
         LOG.info("Killing Hadoop MR job " + this.job.getJobID());
         this.job.killJob();
       }
       this.isCancelled = true;
+      LOG.info("Cancelled job " + this.jobProps.getProperty(ConfigurationKeys.JOB_ID_KEY));
+      if (jobListener != null) {
+        jobListener.onJobCancellation(this.jobState);
+      }
     } catch (IOException ioe) {
       LOG.error("Failed to kill the Hadoop MR job " + this.job.getJobID(), ioe);
       throw new JobException("Failed to kill the Hadoop MR job " + this.job.getJobID(), ioe);
+    } finally {
+      cleanUpWorkingDirectory();
```

Comment:
I think that the problem is that `GobblinOutputCommitter` relies on the data in `mrJobDir`. It uses the files in the folder to create `WorkUnit`, and then cleans up the staging data for each `WorkUnit`. Since the OutputCommitter runs in the Application Master, and Gobblin runs on the Azkaban machine, there could be some race condition which causes the OutputCommitter to fail.